client/session.go

```diff
@@ -417,29 +420,66 @@ func (s *session) Write(namespace string, id string, t time.Time, value float64,
 	w.datapoint.Timestamp = ts
 	w.datapoint.TimestampType = timeType
 	w.datapoint.Annotation = annotation
+
+	wg.Add(1)
+
+	s.RLock()
+	majority = int32(s.majority)
 	w.completionFn = func(result interface{}, err error) {
+		var snapshotSuccess int32
 		if err != nil {
+			atomic.AddInt32(&resultErrs, 1)
 			resultErrLock.Lock()
 			if resultErr == nil {
 				resultErr = err
 			}
-			resultErrs++
 			resultErrLock.Unlock()
+		} else {
+			snapshotSuccess = atomic.AddInt32(&success, 1)
+		}
+		remaining := atomic.AddInt32(&pending, -1)
+		doneAll := remaining == 0
+		switch s.level {
+		case topology.ConsistencyLevelOne:
+			complete := snapshotSuccess > 0 || doneAll
+			if complete && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
+				wg.Done()
+			}
+		case topology.ConsistencyLevelMajority:
+			complete := snapshotSuccess >= majority || doneAll
+			if complete && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
+				wg.Done()
+			}
+		case topology.ConsistencyLevelAll:
+			if doneAll && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
+				wg.Done()
+			}
+		}
+		// Finalize resources that can be returned
+		if doneAll {
+			// Return write to pool
+			s.writeOpPool.Put(w)
```

Comment:
Does this depend on success of write? ie if a write doesn't succeed, we grab a new write from pool for retries?